src/PowerShellEditorServices/Services/DebugAdapter/Handlers/SemanticTokens.cs

```diff
@@ -0,0 +1,145 @@
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models.Proposals;
+
+namespace Microsoft.PowerShell.EditorServices.Handlers
+{
+#pragma warning disable 618
+    internal class SemanticTokens : SemanticTokensHandler
+    {
+        private readonly ILogger _logger;
+        public WorkspaceService _workspaceService;
+        public SemanticTokens(ILogger<SemanticTokens> logger, WorkspaceService workspaceService) : base(new SemanticTokensRegistrationOptions() {
+            DocumentSelector = DocumentSelector.ForLanguage("powershell"),
+            Legend = new SemanticTokensLegend(),
+            DocumentProvider = new Supports<SemanticTokensDocumentProviderOptions>(true,
+                new SemanticTokensDocumentProviderOptions() {
+                    Edits = true
+                }),
+            RangeProvider = true
+        })
+        {
+            _logger = logger;
+            _workspaceService = workspaceService;
+        }
+
+        public override async Task<OmniSharp.Extensions.LanguageServer.Protocol.Models.Proposals.SemanticTokens> Handle(
+            SemanticTokensParams request, CancellationToken cancellationToken)
+        {
+            var result = await base.Handle(request, cancellationToken);
+            return result;
+        }
+
+        public override async Task<SemanticTokensOrSemanticTokensEdits> Handle(SemanticTokensEditsParams request,
+            CancellationToken cancellationToken)
+        {
+            var result = await base.Handle(request, cancellationToken);
+            return result;
+        }
+
+        protected override async Task Tokenize(SemanticTokensBuilder builder, ITextDocumentIdentifierParams identifier,
+            CancellationToken cancellationToken)
+        {
+            ScriptFile file = _workspaceService.GetFile(DocumentUri.GetFileSystemPath(identifier));
+            await Task.Yield();
+            Token[] tokens = file.ScriptTokens;
+            foreach (var token in tokens){
+                pushToken(token, builder);
+            }
+        }
+
+        private static void pushToken(Token token, SemanticTokensBuilder builder){
+            if(token is StringExpandableToken stringExpandableToken){
+                //try parsing tokens within the string
+                if(stringExpandableToken.NestedTokens != null)
+                {
+                    foreach(Token t in stringExpandableToken.NestedTokens){
+                        pushToken(t, builder);
+                    }
+                    return;
+                }
+            }
+
+            //Tokens line and col numbers indexed starting from 1, expecting indexing from 0
+            var line = token.Extent.StartLineNumber - 1;
+            var index = token.Extent.StartColumnNumber - 1;
+
+            builder.Push(line, index, token.Text.Length, MapSemanticToken(token), new string[]{});
```

Comment:
```suggestion
            builder.Push(line, index, token.Text.Length, MapSemanticToken(token), Array.Empty<string>());
```

Reply:
this is more performant because `Array.Empty` doesn't allocate a new array.